tests/unauthenticated_stream.rs

```diff
@@ -0,0 +1,264 @@
+// OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+// CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+#![forbid(
+    anonymous_parameters,
+    box_pointers,
+    legacy_directory_ownership,
+    missing_copy_implementations,
+    missing_debug_implementations,
+    missing_docs,
+    trivial_casts,
+    trivial_numeric_casts,
+    unsafe_code,
+    unstable_features,
+    unused_extern_crates,
+    unused_import_braces,
+    unused_qualifications,
+    unused_results,
+    variant_size_differences,
+    warnings,
+)]
+
+extern crate ring;
+
+use ring::{error, test, unauthenticated_stream};
+// use std::vec::Vec;
+
+#[test]
+fn unauthenticated_stream_chacha20() {
+    test_stream(&unauthenticated_stream::CHACHA20,
+                    "tests/stream_chacha20_tests.txt");
+}
+
+
+fn test_stream(stream_alg: &'static unauthenticated_stream::Algorithm,
+                                                            file_path: &str) {
+    test_stream_key_sizes(stream_alg);
+    test_stream_nonce_sizes(stream_alg).unwrap();
+
+    test::from_file(file_path, |section, test_case| {
+        assert_eq!(section, "");
+        let key_bytes = test_case.consume_bytes("KEY");
+        let nonce = test_case.consume_bytes("NONCE");
+        let plaintext = test_case.consume_bytes("IN");
+        let ct = test_case.consume_bytes("CT");
+        let error = test_case.consume_optional_string("FAILS");
+
+        let mut s_in_out = plaintext.clone();
+        let s_key = unauthenticated_stream::EncryptingKey::new(stream_alg,
+                                                               &key_bytes[..])?;
+
+        let s_result =
+            unauthenticated_stream::encrypt_in_place(&s_key, &nonce[..],
+                                                     &mut s_in_out[..]);
+
+        let mut o_in_out = vec![123u8; 4096];
+        o_in_out.truncate(0);
+        o_in_out.extend_from_slice(&ct[..]);
+        let o_key = unauthenticated_stream::DecryptingKey::new(stream_alg,
+                                                              &key_bytes[..])?;
+
+        let o_result =
+            unauthenticated_stream::decrypt_in_place(&o_key, &nonce[..],
+                                                      &mut o_in_out);
+
+        match error {
+            None => {
+                assert_eq!(Ok(ct.len()), s_result);
+                assert_eq!(&ct[..], &s_in_out[..ct.len()]);
+                assert_eq!(&plaintext[..], o_result.unwrap());
+            },
+            Some(ref error) if error == "WRONG_NONCE_LENGTH" => {
+                assert_eq!(Err(error::Unspecified), s_result);
+                assert_eq!(Err(error::Unspecified), o_result);
+            },
+            Some(error) => {
+                unreachable!("Unexpected error test case: {}", error);
+            },
+        };
+
+        Ok(())
+    });
+}
+
+fn test_stream_key_sizes(stream_alg: &'static unauthenticated_stream::Algorithm) {
```

Comment:
Let's find a way to abstract this test so that the common logic between AEAD and unauthenticated cases can be shared.

Reply:
Done (using a macro, not sure if there's a better way...).